client/my-sites/plans-v2/details.tsx

```diff
@@ -2,20 +2,155 @@
+import { isProductsListFetching } from 'state/products-list/selectors/is-products-list-fetching';
+import { getProductCost } from 'state/products-list/selectors/get-product-cost';
+import { getSelectedSiteSlug } from 'state/ui/selectors';
+
+/**
+ * Type dependencies
+ */
+import type { DetailsPageProps, PurchaseCallback, SelectorProduct } from './types';
+
+import './style.scss';
+
+type Cost = {
+	originalPrice: number;
+	discountedPrice?: number;
+};
+
+const ProductComponent = ( {
+	productSlug,
+	onClick,
+	currencyCode,
+}: {
+	productSlug: string;
+	onClick: PurchaseCallback;
+	currencyCode: string;
+} ) => {
+	const availableProducts = useSelector( ( state ) => getAvailableProductsList( state ) );
+	const price = useSelector( ( state ) => getProductCost( state, productSlug ) ) || 0;
+	const product = slugToSelectorProduct( productSlug );
+	if ( ! product ) {
+		// Exit if invalid product.
+		return null;
+	}
+	const isBundle = [ OPTIONS_JETPACK_SECURITY, OPTIONS_JETPACK_SECURITY_MONTHLY ].includes(
+		productSlug
+	);
+	const JetpackCard = isBundle ? JetpackBundleCard : JetpackProductCard;
+
+	// Gets the cost of the product/plan.
+	// TODO: Pull this all from the API.
+	let cost: Cost = { originalPrice: 0 };
+	if ( isBundle ) {
+		cost = { originalPrice: price || 0 };
+	} else {
+		const productPrices = getProductPrices( product, availableProducts );
+		cost = {
+			originalPrice: productPrices.cost || 0,
+			discountedPrice: productPrices.discountCost,
+		};
+	}
+
+	return (
+		<JetpackCard
+			key={ productSlug }
+			iconSlug={ product.iconSlug }
+			productName={ product.displayName }
+			subheadline={ product.tagline }
+			description={ product.description }
+			currencyCode={ currencyCode }
+			billingTimeFrame={ durationToText( product.term ) }
+			buttonLabel={ productButtonLabel( product ) }
+			onButtonClick={ () => onClick( product ) }
+			features={ { items: [] } }
+			className="details__column"
+			{ ...cost }
+		/>
+	);
+};
+
+const DetailsPage = ( { duration, productSlug, rootUrl }: DetailsPageProps ) => {
+	const siteSlug = useSelector( ( state ) => getSelectedSiteSlug( state ) ) || '';
+	const currencyCode = useSelector( ( state ) => getCurrentUserCurrencyCode( state ) );
+	const isFetchingProducts = useSelector( ( state ) => isProductsListFetching( state ) );
+	const translate = useTranslate();
+	const root = rootUrl.replace( ':site', siteSlug );
+	const durationSuffix = duration ? '/' + durationToString( duration ) : '';
+
+	// If the product slug isn't one that has options, proceed to the upsell.
+	if ( ! ( PRODUCTS_WITH_OPTIONS as readonly string[] ).includes( productSlug ) ) {
+		page.redirect( `${ root }/${ productSlug }${ durationSuffix }` );
+		return null;
+	}
+
+	// If the product is not valid, send the user to the selector page.
+	const product = slugToSelectorProduct( productSlug );
+	if ( ! product ) {
+		page.redirect( root + durationSuffix );
```

Comment:
I wonder if we could use the same check at the controller level for every page to avoid having to validate if the productSlug is valid in each page.